### ISSUE TITLE
Language Server for VS Code and others

### ISSUE THREAD

**Issue by processing-bot** [processing/processing4#117]
In GitLab by @austin.ce on Feb 25, 2018, 21:47

 

## Description

I'm wondering if there has been any work to create a Language Server for Processing yet. It would be very useful for people who would like to use the language outside of the Processing IDE like [this project for VS Code](https://github.com/TobiahZ/processing-vscode). I'm not sure if I'll have time I'd love to help try to get it started if it hasn't been before!

**Comment by processing-bot** [comment]
*Created by: benfry*

Not that I know of; please let us know if you make some progress. I don't know how many people are using VS Code, but perhaps it'd be helpful? Thanks for checking in.

**Comment by processing-bot** [comment]
*Created by: JakubValtar*

I'm working on a new preprocessor which would allow us to run Java 8. Until then I don't think it's possible to create a language server because our current preprocessor is quite picky about which syntax it will accept (it's something between Java 5 and 8) and it's a bit of a mess.

When we are on Java 8 we will need a new autocompletion engine. Extending Java Language Server and possibly moving more functionality from the PDE there seems to be a good option, but we still have to decide on the architecture (do we use LS, what needs to stay in the PDE vs what can be moved/duplicated to the language server and therefore reusable in other editors).

**Comment by processing-bot** [comment]
In GitLab by @austin.ce on Feb 27, 2018, 16:57

Ok, sounds good @JakubValtar. Do you need any help with the preprocessor?

**Comment by processing-bot** [comment]
*Created by: neilcsmith-net*

Just linking two conversations together! :wink: https://forum.processing.org/two/discussion/26740/proposal-development-environment-javascript-prototype

Out of interest, why was the title changed to "for VS Code"?

**Comment by processing-bot** [comment]
In GitLab by @pr1metine on May 28, 2020, 14:50

Any progress? What about [this](https://github.com/processing-language-server/LS4P)? Can't seem to get it working...

**Comment by processing-bot** [comment]
*Created by: jeremydouglass*

Yes, LS4P was a GSOC 2019 project. Here is the final report:

-  https://syamsundarkirubakaran.github.io/lspfinal
-  https://medium.com/processing-foundation/google-summer-of-code-2019-wrap-up-post-3478323bb0ea

Good point that once that is complete then the issue here could probably be closed.

It seems like you opened an issue there about you having problems building it on a windows hyper-v...?

https://github.com/processing-language-server/LS4P/issues/6

**Comment by processing-bot** [comment]
In GitLab by @pr1metine on May 29, 2020, 06:03

Yeah, I was unable to run LS4P on a Ubuntu VM and WSL, let alone on native Windows. Did you, jeremydouglass, or anyone else manage to get it up and running? If that's the case I'll know that it's an issue on my part.

**Comment by processing-bot** [comment]
*Created by: benfry*

We'd still like to see LSP support, though I don't have time to build it myself. [Theia](https://github.com/eclipse-theia/theia) also looks promising if we were to replace the PDE entirely.

**Comment by processing-bot** [comment]
*Created by: benfry*

(Moving this to the Processing 4 repo since we wouldn't be coming back to 3.x to do this.)

**Comment by processing-bot** [comment]
In GitLab by @pr1metine on Aug 15, 2020, 15:34

I am working on a LS. It's in a very early stage though and it can merely format PDE files.

**Comment by processing-bot** [comment]
*Created by: sampottinger*

That's awesome @pr1metine! Obviously I'm sure you are aware but Processing 4 does have a new preprocessor with java 11 syntax support. LMK if you need help wading into any of that!

**Comment by processing-bot** [comment]
In GitLab by @pr1metine on Aug 16, 2020, 02:39

Thank you, @sampottinger ! I'd be happy for your help especially on the Preprocessor and Linting. [This](https://github.com/pr1metine/ls4pde) is what I've been working on so far.

For further questions hit me up in my [Twitter DMs](https://twitter.com/pr1metine).

**Comment by processing-bot** [comment]
*Created by: Cyfine*

I would like to see an official language support VS Code plugin for Processing in the future.

**Comment by processing-bot** [comment]
*Created by: benfry*

Me too! Just need someone to do the work.

**Comment by processing-bot** [comment]
*Created by: kgtkr*

I started this work a few days ago. It is implemented based on the code of Processing Development Environment. This is very unstable, but I have implemented formatters and error reporting.

![image](https://gitlab.com/processing-foundation/processing4/uploads/5e5642406748ab3d92df41ab3ad986f9/139575101-16460199-21fd-497a-a0c1-5b616b70884f.png)


https://github.com/kgtkr/processing-language-server
https://github.com/kgtkr/processing-language-server-vscode

postscript(2021/11/01)

![image](https://gitlab.com/processing-foundation/processing4/uploads/ab4f0c00307ed19caf882a21b0b92fb4/139665455-6c6eeada-a08f-4936-86be-0aaf22bc2198.png)

**Comment by processing-bot** [comment]
*Created by: SableRaf*

Bumping this issue now that Processing 4.0 has been released. 

This would be a great time to revive community efforts towards a Language Server for Processing!

_For anyone interested to learn about the Language Server Protocol (and maybe help with development) here are some useful references:_
- https://microsoft.github.io/language-server-protocol/
- https://code.visualstudio.com/api/language-extensions/language-server-extension-guide

**Comment by processing-bot** [comment]
*Created by: SableRaf*

Note: there was also some work done by @SyamSundarKirubakaran in 2021
The code can be found here: https://github.com/processing-language-server/LS4P and [a conversation about it in the Processing forum](https://discourse.processing.org/t/processing-language-server-gsoc19/9554).

**Comment by processing-bot** [comment]
*Created by: kgtkr*

I was able to easily implement lsp by using the source code of the integrated editor.
However, this is easier to maintain if integrated into processing
Do you have a plan? i might be able to implement

**Comment by processing-bot** [comment]
*Created by: SableRaf*

Hi @kgtkr 👋 Could you describe what you specifically need in order to integrate the LSP in Processing ?

**Comment by processing-bot** [comment]
*Created by: kgtkr*

Once the prototype is complete, please do a code review. Please wait until my work is finished.

**Comment by processing-bot** [comment]
*Created by: kgtkr*

@SableRaf #564

**Comment by processing-bot** [comment]
*Created by: benfry*

Initial work from @kgtkr now incorporated for Processing 4.1. I'm closing this issue since the base support is now present, and we can file separate/more specific issues as needed for adding other enhancements. 

For instance, we don't have an official _client_ yet, but hopefully the community can help fill that hole for various platforms for at least the time being.

**Comment by processing-bot** [comment]
*Created by: github-actions[bot]*

This issue has been automatically locked. To avoid confusion with reports that have already been resolved, closed issues are automatically locked 30 days after the last comment. Please open a new issue for related bugs.
